beater/java_attacher/java_attacher.go

```diff
@@ -25,50 +25,369 @@ import (
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
+	"runtime"
 	"strings"
-
-	"github.com/elastic/elastic-agent-libs/logp"
+	"sync"
+	"time"
 
 	"github.com/elastic/apm-server/beater/config"
+	"github.com/elastic/elastic-agent-libs/logp"
 )
 
+var encounteredJvmCache = make(map[string]*JvmDetails)
+
+// javaAttacher is bundled by the server
+var javaAttacher = filepath.FromSlash("./java-attacher.jar")
+
+type JvmDetails struct {
+	user        string
+	uid         string
+	gid         string
+	pid         string
+	startTime   string
+	command     string
+	version     string
+	cmdLineArgs string
+}
+
 type JavaAttacher struct {
-	cfg    config.JavaAttacherConfig
-	logger *logp.Logger
+	logger               *logp.Logger
+	enabled              bool
+	discoveryRules       []discoveryRule
+	rawDiscoveryRules    []map[string]string
+	agentConfigs         map[string]string
+	downloadAgentVersion string
+	javaBin              string
+}
+
+func (j *JavaAttacher) addDiscoveryRule(rule discoveryRule) {
+	j.discoveryRules = append(j.discoveryRules, rule)
+	j.logger.Debugf("added discovery rule: %v", rule.asString())
 }
 
 func New(cfg config.JavaAttacherConfig) (JavaAttacher, error) {
-	if cfg.JavaBin == "" {
+	logger := logp.NewLogger("java-attacher")
+	if _, err := os.Stat(javaAttacher); err != nil {
+		return JavaAttacher{}, err
+	}
+	attacher := JavaAttacher{
+		logger:               logger,
+		enabled:              cfg.Enabled,
+		agentConfigs:         cfg.Config,
+		downloadAgentVersion: cfg.DownloadAgentVersion,
+		rawDiscoveryRules:    cfg.DiscoveryRules,
+		javaBin:              cfg.JavaBin,
+	}
+	for _, flag := range cfg.DiscoveryRules {
+		for name, value := range flag {
+			switch name {
+			case "include-all":
+				attacher.addDiscoveryRule(includeAllRule{})
+			case "include-user":
+				attacher.addDiscoveryRule(userDiscoveryRule{user: value, isIncludeRule: true})
+			case "exclude-user":
+				attacher.addDiscoveryRule(userDiscoveryRule{user: value, isIncludeRule: false})
+			case "include-main":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-main")
+			case "exclude-main":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			case "include-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-vmarg")
+			case "exclude-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			default:
+				logger.Errorf("Unknown discovery rule - '%v'", name)
+			}
+		}
+	}
+	return attacher, nil
```

Comment:
**_question_**: does it make sense to return an error if there are no (known) discovery rules ?

Reply:
I suppose we can return `nil` in this case